stonybook/pipeline/episode_break_prominence/add_break_prominence.py

```diff
@@ -0,0 +1,189 @@
+                                                sent_lemma[key].append(entityLemma.attrib['lemma'])
+
+                                lemmas.update(sent_lemma)
+                        para_breaks[int(paragraphs.attrib['num'])] = key
+
+                if 'desc' in cur.keys():
+                    chapterCounter += 1
+                    chapter_breaks[chapterCounter] = [int(paragraphs.attrib['num']), key]
+
+    with open(os.path.join(output_dir, 'lemmas_pickle.pkl'), 'wb') as f:
+        pickle.dump(lemmas, f) 
+
+    return para_breaks, chapter_breaks
+
+def build_graph(lemma_dict, para_breaks, N = 150):
+    edges = list()
+    para_breakSentences = [x for x in para_breaks.values()]
+    if len(lemma_dict.keys()) == 0:
+        return edges
+    for idx in range(max(lemma_dict.keys()) + 1):
+        if idx not in para_breakSentences:
+            continue
+        for n in range(-N, 1 + N):
+            if idx + n not in lemma_dict or idx == n:
+                continue
+
+            common_lemmas = lemma_dict[idx].intersection(lemma_dict[idx + n])
+            new_common_lemmas = set()
+            for x in common_lemmas:
+                if x not in "!\"#$%&'()*+, -./:;<=>?@[\]^_`{|}~":
+                    new_common_lemmas.add(x)
+            common_lemmas = new_common_lemmas
+            for i in range(len(common_lemmas)):
+                edges.append([idx, idx + n])
+    return edges
+
+#normalize densities in range 1-100
+def normalize_density(densities, para_breaks, max_sent_num):
+    para_breakSentences = [x for x in para_breaks.values()]
+    numParas = len(para_breakSentences)
+    densities = [densities[x] for x in range(max_sent_num+1) if x in para_breakSentences]
+
+    density = {x:0 for x in range(0, numParas)}
+    try:
+        maxDensity = max(densities)
+        minDensity = min(densities)
+    except ValueError:
+        maxDensity = 0
+        minDensity = 0
+    rangeofDensities = maxDensity - minDensity
+    for i in range(numParas):
+        try:
+            density[i] = ((densities[i] - minDensity)*100) / rangeofDensities
+        except ZeroDivisionError:
+            density[i] = 100
+    return dict(density)
+
+#normalize prominences to range 0-1
+def normalize_prominences(peaks, prominences):
+    normalized_prominence = {}
+    l = len(peaks)
+    print(peaks, prominences)
+    try:
+        maxVal = max(prominences)
+        minVal = min(prominences)
+    except:
+        maxVal = 0
+        minVal = 0
+    rangeVals = maxVal - minVal
+    for i in range(l):
+        try:
+            normalized_prominence[peaks[i]] = round((prominences[i] - minVal) / rangeVals, 2)
+        except ZeroDivisionError:
+            continue
+    
+    return normalized_prominence
+
+#get densities using inverse exponential weighted function
+def get_densities(edges, para_breaks, max_sent_num):
+    density = [0 for _ in range(max_sent_num+1)]
+    for x, y in edges:
+        for i in range(x, y):
```

Comment:
Seems like a bug. The values of y can be smaller than x, in which case this code will not run.